files/scripts/swss.sh

```diff
@@ -107,6 +107,100 @@ function clean_up_tables()
     end" 0
 }
 
+# This function cleans up the chassis db table entries created ONLY by this asic
+# This is used to do the clean up operation when the line card / asic reboots
+# When the asic/lc is RE-booting, the chassis db server is supposed to be running 
+# in the supervisor.  So the clean up is done when only the chassis db connectable. 
+# Otherwise no need to do the clean up since both the supervisor and line card may be 
+# rebooting (the whole chassis scenario)
+# The clean up operation is required to delete only those entries created by
+# the asic that is rebooted. Entries from the following tables are deleted in the order
+# given below
+#   (1) SYSTEM_NEIGH
+#   (2) SYSTEM_INTERFACE
+#   (3) SYSTEM_LAG_MEMBER_TABLE
+#   (4) SYSTEM_LAG_TABLE
+#   (5) The corresponding LAG IDs of the entries from SYSTEM_LAG_TABLE
+#       SYSTEM_LAG_ID_TABLE and SYSTEM_LAG_ID_SET are adjusted appropriately
+function clean_up_chassis_db_tables()
+{
+    if [[ !($($SONIC_DB_CLI CHASSIS_APP_DB PING | grep -c True) -gt 0) ]]; then
+        return
+    fi
+
+    lc=`$SONIC_DB_CLI CONFIG_DB  hget 'DEVICE_METADATA|localhost' 'hostname'`
+    asic=`$SONIC_DB_CLI CONFIG_DB  hget 'DEVICE_METADATA|localhost' 'asic_name'`
+
+    # First, delete SYSTEM_NEIGH entries
+    $SONIC_DB_CLI CHASSIS_APP_DB EVAL "
+    local host = string.gsub(ARGV[1], '%-', '%%-')
+    local dev = ARGV[2]
+    local ps = 'SYSTEM_NEIGH*|' .. host .. '|' .. dev
+    local keylist = redis.call('KEYS', 'SYSTEM_NEIGH*')
+    for j,key in ipairs(keylist) do
+        if string.match(key, ps) ~= nil then
+            redis.call('DEL', key)
+        end
+    end
+    return " 0 $lc $asic
+
+    # Wait for some time before deleting system interface so that the system interface's "object in use"
+    # is cleared in both orchangent and in syncd. Without this delay, the orchagent clears the refcount
+    # but the syncd (meta) still has no-zero refcount. Because of this, orchagent gets "object still in use"
+    # error and aborts.
+
+    sleep 30
```

Comment:
Can we also add a check to see if the number of SYSTEM_NEIGHBOR entries ( and similarly below for SYSTEM_LAG_MEMBER_TABLE) **are non-zero** before adding this sleep(). In case of retries when swss systemd service restarts multiple times etc we need not each time wait for 45 sec.

Reply:
The problem is not with number of system neighbors or system lag members in the chassis app db. The problem is with timing in processing in other remote asics and linecards. It depends on how much time the remote asics/linecards take to remove the referenced entries in that asic/linecard.

Reply:
That is true, I was pointing to a few cases as below where we can skip the sleep altogether 

(a) no SYSTEM_NEIGHBORS/LAG for that asic/LC ( so ideally nothing to clean locally and remote)
or in scenario below 
(b) where swss restarts, we check and cleanup SYSTEM_NEIGHBORS for that asic/LC in chassis_db. Wait for 30 secs and make sure that the processing in remote asic/linecard is done.
   > But not for some reason swss restarts again (assume orchgent crashed). Again we will do same cleanup SYSTEM_NEIGHBORS for that asic/LC in chassis_db -- but this time there are no entries as in last round we cleaned up all .
  > So now there is no need to wait for 30 sec and 15 sec as there is no SYSTEM_NEIGHBOR/LAG entries relating to my LC/asic present in CHASSIS_DB

But don't block merge for this comment -- we can fine tune this in subsequent commits too

Reply:
Thanks for the clarification. I'll fix this.